pyodide-build/pyodide_build/config.py

```diff
@@ -0,0 +1,182 @@
+
+from .common import _environment_substitute_str, exit_with_stdio
+from .logger import logger
+
+
+class ConfigManager:
+    """
+    Configuration manager for Package build process.
+
+    The configuration manager is responsible for loading configuration from various sources.
+    The configuration can be loaded from the following sources (in order of precedence):
+
+        1. Command line arguments
+        2. Environment variables
+        3. Configuration file
+        4. Makefile.envs
+        5. Default values
+    """
+
+    def __init__(self, pyodide_root: Path):
+        self.pyodide_root = pyodide_root
+        self._config = {
+            **self._load_default_config(),
+            **self._load_makefile_envs(),
+            **self._load_config_file(),
+            **self._load_config_from_env(dict(os.environ)),
+        }
+
+    def _load_default_config(self) -> dict[str, str]:
+        return {
+            k: _environment_substitute_str(
+                v, env={"PYODIDE_ROOT": str(self.pyodide_root)}
+            )
+            for k, v in DEFAULT_CONFIG.items()
+        }
+
+    def _load_makefile_envs(self) -> dict[str, str]:
+        makefile_vars = self._get_make_environment_vars()
+        computed_vars = {
+            k: _environment_substitute_str(v, env=makefile_vars)
+            for k, v in DEFAULT_CONFIG_COMPUTED.items()
+        }
+
+        return {
+            BUILD_VAR_TO_KEY[k]: v
+            for k, v in makefile_vars.items()
+            if k in BUILD_VAR_TO_KEY
+        } | computed_vars
+
+    def _get_make_environment_vars(self) -> dict[str, str]:
+        """
+        Load environment variables from Makefile.envs
+        """
+        environment = {}
+        result = subprocess.run(
+            ["make", "-f", str(self.pyodide_root / "Makefile.envs"), ".output_vars"],
+            capture_output=True,
+            text=True,
+            env={"PYODIDE_ROOT": str(self.pyodide_root)},
+        )
+
+        if result.returncode != 0:
+            logger.error(
+                "ERROR: Failed to load environment variables from Makefile.envs"
+            )
+            exit_with_stdio(result)
+
+        for line in result.stdout.splitlines():
+            equalPos = line.find("=")
+            if equalPos != -1:
+                varname = line[0:equalPos]
+
+                if varname not in BUILD_VAR_TO_KEY:
+                    continue
+
+                value = line[equalPos + 1 :]
+                value = value.strip("'").strip()
+                environment[varname] = value
+
+        return environment
+
+    def _load_config_from_env(self, env: dict[str, str]) -> dict[str, str]:
+        return {
+            BUILD_VAR_TO_KEY[key]: env[key] for key in env if key in BUILD_VAR_TO_KEY
+        }
+
+    def _load_config_file(self) -> dict[str, str]:
+        # TODO: Implement this
+        return {}
+
+    @property
+    def config(self) -> dict[str, str]:
+        return self._config.copy()
```

Comment:
It isn't really ideal that this copies, for example in `to_env` we make two copies of the dictionary, once to access `self.config`. What about 
```py
from types import MappingProxyType
# ...
    @property
    def config(self) -> dict[str, str]:
        return MappingProxyType(self._config)
```